src/sysfs_fuse.c

```diff
@@ -126,6 +127,93 @@ static int sys_devices_system_cpu_online_read(char *buf, size_t size,
 	return total_len;
 }
 
+static int filler_sys_devices_system_cpu(const char *path, void *buf,
+		fuse_fill_dir_t filler)
+{
+	__do_free char *cg = NULL, *cpuset = NULL;
+	struct fuse_context *fc = fuse_get_context();
+	pid_t initpid;
+
+	initpid = lookup_initpid_in_store(fc->pid);
+
+	if (initpid <= 1 || is_shared_pidns(initpid))
+		initpid = fc->pid;
+	cg = get_pid_cgroup(initpid, "cpuset");
+	if (!cg)
+		return 0; 
+
+	prune_init_slice(cg);
+
+	cpuset = get_cpuset(cg);
+	if (!cpuset)
+		return 0; 
+
+	char *tok;
+	lxc_iterate_parts(tok, cpuset, ",") {
+		if (strstr(tok, "-")) {
+			char *cpu[2];
+			int i = 0;
+
+			char *t;
+			lxc_iterate_parts(t, tok, "-") {
+				if(i > 1)
+					return 0;
+				cpu[i++] = t;
+			}
+
+			for(i=atoi(cpu[0]); i<=atoi(cpu[1]); i++) {
+				char cpu_str[10];
+				sprintf(cpu_str, "cpu%d", i);
+				if (DIR_FILLER(filler, buf, cpu_str, NULL, 0) != 0) {
+					return -ENOENT;
+				}
+			}
+		}
```

Comment:
Hm, I'm not a fan of the `strstr()` based cpuset parsing here. I have logic for this in LXC:
```c
static ssize_t get_max_cpus(char *cpulist)
{
	char *c1, *c2;
	char *maxcpus = cpulist;
	size_t cpus = 0;

	c1 = strrchr(maxcpus, ',');
	if (c1)
		c1++;

	c2 = strrchr(maxcpus, '-');
	if (c2)
		c2++;

	if (!c1 && !c2)
		c1 = maxcpus;
	else if (c1 > c2)
		c2 = c1;
	else if (c1 < c2)
		c1 = c2;
	else if (!c1 && c2)
		c1 = c2;

	errno = 0;
	cpus = strtoul(c1, NULL, 0);
	if (errno != 0)
		return -1;

	return cpus;
}

/* Taken over modified from the kernel sources. */
#define NBITS 32 /* bits in uint32_t */
#define DIV_ROUND_UP(n, d) (((n) + (d)-1) / (d))
#define BITS_TO_LONGS(nr) DIV_ROUND_UP(nr, NBITS)

static void set_bit(unsigned bit, uint32_t *bitarr)
{
	bitarr[bit / NBITS] |= (1 << (bit % NBITS));
}

static void clear_bit(unsigned bit, uint32_t *bitarr)
{
	bitarr[bit / NBITS] &= ~(1 << (bit % NBITS));
}

static bool is_set(unsigned bit, uint32_t *bitarr)
{
	return (bitarr[bit / NBITS] & (1 << (bit % NBITS))) != 0;
}

/* Create cpumask from cpulist aka turn:
 *
 *	0,2-3
 *
 * into bit array
 *
 *	1 0 1 1
 */
static uint32_t *lxc_cpumask(char *buf, size_t nbits)
{
	__do_free uint32_t *bitarr = NULL;
	char *token;
	size_t arrlen;

	arrlen = BITS_TO_LONGS(nbits);
	bitarr = calloc(arrlen, sizeof(uint32_t));
	if (!bitarr)
		return ret_set_errno(NULL, ENOMEM);

	lxc_iterate_parts(token, buf, ",") {
		errno = 0;
		unsigned end, start;
		char *range;

		start = strtoul(token, NULL, 0);
		end = start;
		range = strchr(token, '-');
		if (range)
			end = strtoul(range + 1, NULL, 0);

		if (!(start <= end))
			return ret_set_errno(NULL, EINVAL);

		if (end >= nbits)
			return ret_set_errno(NULL, EINVAL);

		while (start <= end)
			set_bit(start++, bitarr);
	}

	return move_ptr(bitarr);
}


	/* Get maximum number of cpus found in possible cpuset. */
	maxposs = get_max_cpus(cpuset);
	if (maxposs < 0 || maxposs >= INT_MAX - 1)
                /* do something */
	
	possmask = lxc_cpumask(posscpus, maxposs);
	if (!possmask)
                /* do something */
                
	for (size_t i = 0; i <= nbits; i++) {
		if (!is_set(i, bitarr))
			continue;
			
                sprintf(cpu_str, "cpu%d", i);
        }
```